proto/errors.go

```diff
@@ -73,16 +73,30 @@ var (
 type KafkaError struct {
 	errno   int16
 	message string
+	// An additional message explaining cases where the client returns this error internally
+	// instead of directly from a Kafka broker.
+	syntheticMessage string
 }
 
 func (err *KafkaError) Error() string {
+	if err.syntheticMessage != "" {
+		fmt.Sprintf("%s (%d %s)", err.message, err.errno, err.syntheticMessage)
+	}
 	return fmt.Sprintf("%s (%d)", err.message, err.errno)
 }
 
 func (err *KafkaError) Errno() int {
 	return int(err.errno)
 }
 
+func (err *KafkaError) SyntheticError() string {
+	return err.syntheticMessage
+}
+
+func (err *KafkaError) NewSynthetic(message string) *KafkaError {
+	return &KafkaError{err.errno, err.message, message}
```

Comment:
Bah, clearly I didn't think it through either because I fixed up the tests without considering the clients do the same thing.

IMHO it's completely bogus to assume that all errors are singletons and checking pointer equality is not a proper method of error handling. I hope we agree on this to start with?

For a fix, usually it's proper to do a type switch, but I don't actually want to statically declare a new struct for every Kafka protocol error. We can define an Equals method on *KafkaErr and then clients can write `case proto.ErrNotLeaderForPartition.Equals(err)` which will try to type assert `err` into `*KafkaError` internally. It's a little gross but the clients shouldn't really notice.

If we want to be guaranteed we hit every affected site, we could make the declared error vars just the ints (not `*KafkaError`) but then either the clients have to do the type assertion or we have to make a free helper function in `kafka/proto`. I think I'd rather do the former approach but I could be talked out of it.